### DISCUSSION TITLE
operation exchange.declare caused a channel exception access_refused:

### DISCUSSION THREAD

**Discussion by doing-cr7** [rabbitmq/rabbitmq-server#3994] [Questions] [answered]
operation exchange.declare caused a channel exception access_refused:

**Comment by lukebakken** [comment, marked as answer]
The user does not have permissions to declare the exchange. Please read this document:

https://www.rabbitmq.com/access-control.html

You probably need to run this command. Replace `username` with the user account you're using. Note that this grants ALL permissions to the user:

```
rabbitmqctl set_permissions -p "/" "username" ".*" ".*" ".*"
```

**Reply by fusion2222** [reply]
You don't really need to give ALL permissions to any user. Just give user only the permissions he needs.

In your case it seems it needs only permissions to create (or configure, as RabbitMQ calls it) specific exchange. 

```bash
rabbitmqctl set_permissions -p "/" "username" "<name of your exchange>" "<keep this unchanged>" "<keep this unchanged>"
```

**Reply by michaelklishin** [reply]
Luke's goal was to explain what the error says. What permission granularity is appropriate for this environment, we cannot know. Sometimes it's perfectly fine to grant full permissions.

**Reply by fusion2222** [reply]
We actually can - `exchange.declare` specifies, you need to grant declare or "configure" privileges as RabbitMQ calls it. In this case granting read or write privileges is not needed and is not cause of this error.

**Reply by michaelklishin** [reply]
> We actually can - `exchange.declare` specifies, you need to grant declare or "configure" privileges as RabbitMQ calls it. In this case granting read or write privileges is not needed and is not cause of this error.

We don't know enough about the environment and this user'a goals. This kind of tunnel vision can be counterproductive.
